src/mpi/ccl/cclcomm.c

```diff
@@ -0,0 +1,259 @@
+    }
+
+  fn_exit:
+    return mpi_errno;
+  fn_fail:
+    mpi_errno = MPI_ERR_ARG;
+    goto fn_exit;
+}
+
+int MPIR_NCCL_datatype_is_supported(MPI_Datatype dtype)
+{
+    switch (dtype) {
+      case MPI_CHAR:
+      case MPI_SIGNED_CHAR:
+      case MPI_CHARACTER:
+      case MPI_INTEGER1:
+      case MPI_INTEGER:
+      case MPI_INT8_T:
+      case MPI_BYTE:
+      case MPI_UINT8_T:
+      case MPI_INT:
+      case MPI_INT32_T:
+      case MPI_INTEGER4:
+      case MPI_UNSIGNED:
+      case MPI_UINT32_T:
+      case MPI_LONG:
+      case MPI_INT64_T:
+      case MPI_INTEGER8:
+      case MPI_UNSIGNED_LONG:
+      case MPI_UINT64_T:
+      case MPI_DOUBLE_COMPLEX:
+      case MPI_FLOAT:
+      case MPI_C_COMPLEX:
+      case MPI_CXX_FLOAT_COMPLEX:
+      case MPI_DOUBLE:
+      case MPI_C_DOUBLE_COMPLEX:
+      case MPI_COMPLEX8:
+      case MPI_CXX_DOUBLE_COMPLEX:
+        return 1;
+      default:
+        return 0;
+    }
+}
+
+int MPIR_NCCL_get_datatype(MPI_Datatype dtype, ncclDataType_t *nccl_dtype)
+{
+    int mpi_errno = MPI_SUCCESS;
+
+    switch (dtype) {
+      case MPI_CHAR:
+      case MPI_SIGNED_CHAR:
+      case MPI_CHARACTER:
+        *nccl_dtype = ncclChar;
+        break;
+      case MPI_INTEGER1:
+      case MPI_INTEGER:
+      case MPI_INT8_T:
+        *nccl_dtype = ncclInt8;
+        break;
+      case MPI_BYTE:
+      case MPI_UINT8_T:
+        *nccl_dtype = ncclUint8;
+        break;
+      case MPI_INT:
+      case MPI_INT32_T:
+      case MPI_INTEGER4:
+        *nccl_dtype = ncclInt32;
+        break;
+      case MPI_UNSIGNED:
+      case MPI_UINT32_T:
+        *nccl_dtype = ncclUint32;
+        break;
+      case MPI_LONG:
+      case MPI_INT64_T:
+      case MPI_INTEGER8:
+        *nccl_dtype = ncclInt64;
+        break;
+      case MPI_UNSIGNED_LONG:
+      case MPI_UINT64_T:
+        *nccl_dtype = ncclUint64;
+        break;
+      case MPI_DOUBLE_COMPLEX:
+        *nccl_dtype = ncclFloat16;
+      case MPI_FLOAT:
+      case MPI_C_COMPLEX:
+      case MPI_CXX_FLOAT_COMPLEX:
+        *nccl_dtype = ncclFloat32;
+        break;
+      case MPI_DOUBLE:
+      case MPI_C_DOUBLE_COMPLEX:
+      case MPI_COMPLEX8:
+      case MPI_CXX_DOUBLE_COMPLEX:
+        *nccl_dtype = ncclFloat64;
+        break;
+      default:
+        goto fn_fail;
+    }
+
+  fn_exit:
+    return mpi_errno;
+  fn_fail:
+    mpi_errno = MPI_ERR_ARG;
+    goto fn_exit;
+}
+
+int MPIR_NCCL_Allreduce(const void *sendbuf, void *recvbuf, MPI_Aint count, MPI_Datatype datatype,
+                        MPI_Op op, MPIR_Comm * comm_ptr, MPIR_Errflag_t errflag)
+{
+    int mpi_errno = MPI_SUCCESS;
+    cudaError_t ret;
+
+    ncclRedOp_t ncclOp;
+    mpi_errno = MPIR_NCCL_get_red_op(op, &ncclOp);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    ncclDataType_t ncclDatatype;
+    mpi_errno = MPIR_NCCL_get_datatype(datatype, &ncclDatatype);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    if(!comm_ptr->cclcomm) {
+      mpi_errno = MPIR_CCLcomm_init(comm_ptr, comm_ptr->rank);
```

Comment:
setup on first use makes sense. the cleanup call to `MPIR_CCLcomm_free` is missing from this PR at the moment.